lib/Qi.pm

```diff
@@ -10,7 +14,9 @@ sub startup {
   my $config = $self->plugin('Config');
 
   #Model
-   $self->helper(pg => sub { state $pg = Mojo::Pg->new(shift->config('pg')) });
+  $self->helper(pg => sub { state $pg = Mojo::Pg->new(shift->config('pg')) });
+  $self->helper(sql => sub{ state $sql = SQL::Abstract->new});
```

Comment:
This is an excellent thought process, and you are absolutely not wrong to think this way.  In fact, continue to always think this way!  This is a difficult problem to get right.  There is no absolute right or wrong answer.  The way to go about this is to consider the situation carefully, and it is very important to understand the details and consequences (good and bad) of your decisions.

So, what we're talking about here...  This is a *design* question.  I think, from all that I've been reading in the past few years, that you should in general design for readability, usability, ease, etc...  In general, design patters are to help developers, not computers.  I think the trouble with so much in software development is premature optimization.  Everyone wants to think about how their product is going to be the next Facebook, and it'll have at the moment of launch.  Reality doesn't work like that.  Cross that bridge when you get to it...  For starters, optimize for readability and usability.

Along these lines, now, here's the issue with the design patter you have chosen.  (Again, it's *very* good you're thinking this way, just be very aware of the details.)  You have opted to create the SQL::Abstract object once.  What's the cost of creating once for each model?  How many models do you have?  Four?  Five?  Twenty?  What's the impact of creating an SQL::Abstract object?  I honestly do not know the answer to this whatsoever.  I may be wrong, but I do sincerely believe that it is so low, so low as to be negligible.  Perhaps it may be non-negligible if the number were millions.  But, again, I really do not know.  So, before you pre-optimize, understand what you're optimizing for.  Know what your gains will be.  If you really suspect that creating an SQL::Abstract object for each model may be non-negligible, you would be extremely encouraged to benchmark it.  It's not that difficult really.  First, you have tests (You do have tests, right??  ;)  Next you time your tests.  Finally, you run your tests against the two opposing design patterns.  Now, what's the difference?  It would probably be wise to have a hypothesis in advance, or a target or whatever that defines what negligible is, and if the tests show that the optimized pattern is negligible by your definition, then the benefit is none and therefore the optimization is unnecessary.

Now, why *not* optimized if it's perfectly easy to do so, even, on the face of it, easier?!  As long as the optimization doesn't make it worse, and probably makes it better, why not optimize??  For the sake of what your primary design patterns should be optimizing for: not performance, but readability, maintainability, usability...

Why is this "optimized" design pattern less ideal from a developer's point of view?  The reason is that you *want* your modules to be reusable, you want them to be usable outside of this one application.  You want your modules to be self-contained.  If you know that your module is going to need a new SQL::Abstract object, you might as well allow it to create it on its own.  Otherwise, all subsequent applications that wish to use your module will be required to provide this object to your module.  These types of requirements can really add up.

Sometimes, however, you do want your outside application to provide an object to your module.  Why?  How to determine when should the module create its own object and when it should require an object be passed to it?  Pretty easy, for the most part...  When an application needs to provide *instance* data, you pass the object that the application has been using.  This is why the `pg` object is created in the application.  You want the application to connect to the database and then provide that connection to the modules for use.  There is information within an object for the application to provide to the modules.  The SQL::Abstract object has no data to share between modules.  Each module has its own purpose for the SQL::Abstract object, with nothing to share.  You *could* require that each model module connects independently to the database, but that is a known optimization that is absolutely non-negligible.  And it's unnecessary to require each to do so.  Each needs to connect to the same database, so you might as well share that information.  But in the case of SQL::Abstract, there is nothing to be shared, aside from the object initialization itself (the negligible part).

Hoo, that was a lot to say.  But your question and this response exactly touched on some of the most exciting parts of this PR, what makes this PR your next turning point in understanding and advancement in this field.  I feel like I kind of trailed off in clarity towards the end, but I think in general I captured well the primary things I wished to get across to you.  Please let me know if you have any questions about this.  Lots of discussion could be had here.